guppylang/prelude/builtins.py

```diff
@@ -559,6 +561,27 @@ def __len__(self: array[L, n]) -> int: ...
     def __new__(): ...
 
 
+@guppy.extend_type(sized_iter_type_def)
+class SizedIter:
+    """A wrapper around an iterator type `T` promising that the iterator will yield
+    exactly `n` values.
+
+    Annotating an iterator with an incorrect size is undefined behaviour.
+    """
+
+    @guppy.custom(NoopCompiler())
+    def __new__(iterator: L @ owned) -> "SizedIter[L, n]":  # type: ignore[type-arg]
+        """Casts an iterator into a `SizedIter`."""
+
+    @guppy.custom(NoopCompiler())
+    def unwrap_iter(self: "SizedIter[L, n]" @ owned) -> L:
+        """Extracts the actual iterator."""
+
+    @guppy.custom(NoopCompiler())
+    def __iter__(self: "SizedIter[L, n]" @ owned) -> L:
+        """Extracts the actual iterator."""
```

Comment:
what's the difference between these? (is there one?)

Reply:
It's the same, I thought we might want to have a non-dunder method to get the wrapped iterator?